src/dmd/must_use.d

```diff
@@ -0,0 +1,198 @@
+/**
+ * Compile-time checks associated with the @mustUse attribute.
+ *
+ * Copyright: Copyright (C) 1999-2022 by The D Language Foundation, All Rights Reserved
+ * License:   $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:    $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/must_use.d, _must_use.d)
+ * Documentation:  https://dlang.org/phobos/dmd_must_use.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/must_use.d
+ */
+
+module dmd.must_use;
+
+import dmd.dscope;
+import dmd.dsymbol;
+import dmd.expression;
+
+/**
+ * Check whether discarding an expression would violate the requirements of
+ * @mustUse. If so, emit an error.
+ *
+ * Params:
+ *   e = the expression to check
+ *   sc = scope in which `e` was semantically analyzed
+ *
+ * Returns: true on error, false on success.
+ */
+bool checkMustUse(Expression e, Scope* sc)
+{
+    import dmd.id : Id;
+
+    if (!isAssignment(e) && !isIncrementOrDecrement(e))
+    {
+        assert(e.type);
+        auto sym = e.type.toDsymbol(sc);
+        auto sd = sym ? sym.isStructDeclaration() : null;
+        // isStructDeclaration returns non-null for both structs and unions
+        if (sd && hasMustUseAttribute(sd, sc))
+        {
+            e.error("ignored value of `@%s` type `%s`; prepend a `cast(void)` if intentional",
+                Id.udaMustUse.toChars(), e.type.toPrettyChars(true));
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * Called from a symbol's semantic to check for reserved usage of @mustUse.
+ *
+ * If such usage is found, emits an errror.
+ *
+ * Params:
+ *   sym = symbol to check
+ */
+void checkMustUseReserved(Dsymbol sym)
+{
+    import dmd.errors : error;
+    import dmd.id : Id;
+
+    if (sym.userAttribDecl is null || sym.userAttribDecl.atts is null)
+        return;
+
+    // Can't use foreachUda (and by extension hasMustUseAttribute) while
+    // semantic analysis of `sym` is still in progress
+    // TODO: factor out common code from this function and checkGNUABITag
+    foreach (exp; *sym.userAttribDecl.atts)
+    {
+        if (isMustUseAttribute(exp))
+        {
+            if (sym.isFuncDeclaration())
+            {
+                error(sym.loc, "`@%s` on functions is reserved for future use",
+                    Id.udaMustUse.toChars());
+                sym.errors = true;
+            }
+            else if (sym.isClassDeclaration() || sym.isEnumDeclaration())
+            {
+                error(sym.loc, "`@%s` on `%s` types is reserved for future use",
+                    Id.udaMustUse.toChars(), sym.kind());
+                sym.errors = true;
+            }
+        }
+    }
+}
+
+/**
+ * Returns true if the given expression is an assignment, either simple (a = b)
```

Comment:
Use DDoc style comments, such as `Returns:` note the :

Reply:
http://dtest.dlang.io/artifact/website-b0500bb994fbd83fc3e6192d631db59bd84ec19a-52980191da4ba5776ce3b3738b0dc5cf/web/library-prerelease/dmd/must_use/check_must_use.html

also present in generated docs, although I didn't realise it would be.